pkg/kv/kvserver/loqrecovery/storage.go

```diff
@@ -0,0 +1,187 @@
+			CreationTime: created,
+		}, nil
+	}
+	return StoredPlan{}, nil
+}
+
+func (s PlanStorage) RemovePlans() error {
+	return s.visitPlanFiles(func(info fs.FileInfo) error {
+		return s.fs.Remove(filepath.Join(s.path, info.Name()))
+	})
+}
+
+func (s PlanStorage) SavePlan(plan loqrecoverypb.ReplicaUpdatePlan) error {
+	existing, err := s.HasPlan()
+	if err != nil {
+		return errors.Wrap(err, "failed to perform existing recovery plan check")
+	}
+	if !existing.Empty() {
+		return errors.Newf("can not set recovery plan %s, plan file %s is already set", plan.PlanID,
+			existing.Name)
+	}
+
+	if err := s.ensureStoreExists(); err != nil {
+		return err
+	}
+
+	planBaseName := s.planFullName(plan.PlanID.String())
+	tmpFileName := planBaseName + ".tmp"
+	defer func() { _ = s.fs.Remove(tmpFileName) }()
+
+	if err := func() error {
+		outFile, err := s.fs.Create(tmpFileName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create file %q", tmpFileName)
+		}
+		defer func() { _ = outFile.Close() }()
+		jsonpb := protoutil.JSONPb{Indent: "  "}
+		var out []byte
+		if out, err = jsonpb.Marshal(plan); err != nil {
+			return errors.Wrap(err, "failed to marshal recovery plan")
+		}
+		if _, err = outFile.Write(out); err != nil {
+			return errors.Wrap(err, "failed to write recovery plan")
+		}
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	if err := s.fs.Rename(tmpFileName, planBaseName+".json"); err != nil {
+		_ = s.fs.Remove(tmpFileName)
+		return errors.Wrap(err, "failed to rename temp plan file")
+	}
+	return nil
+}
+
+func (s PlanStorage) LoadPlan(plan StoredPlan) (loqrecoverypb.ReplicaUpdatePlan, error) {
+	f, err := s.fs.Open(plan.Name)
+	if err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err, "failed to open plan file %q",
+			plan.Name)
+	}
+	defer func() { _ = f.Close() }()
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err, "failed to read plan file %q",
+			plan.Name)
+	}
+
+	var nodeUpdates loqrecoverypb.ReplicaUpdatePlan
+	jsonpb := protoutil.JSONPb{Indent: "  "}
+	if err = jsonpb.Unmarshal(data, &nodeUpdates); err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err,
+			"failed to unmarshal plan from file %q", plan.Name)
+	}
+
+	if s.planFullName(nodeUpdates.PlanID.String()+".json") != plan.Name {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Newf("loaded plan has id %s while filename is %s, plan id must match filename without extension",
+			nodeUpdates.PlanID, plan.Name)
+	}
+
+	return nodeUpdates, nil
+}
+
+func (s PlanStorage) visitPlanFiles(visitor func(fs.FileInfo) error) error {
+	names, err := s.fs.List(s.path)
+	if err != nil {
+		// Error means we can't access recovery dir. That means no plans are
+		// available.
+		return nil //nolint:returnerrcheck
+	}
+	filePattern := regexp.MustCompile(`[a-zA-Z0-9-]*\.json`)
+	for _, n := range names {
+		if filePattern.MatchString(n) {
```

Comment:
Do we need this validation? Can't we assume any file here is a plan file, or at least any file with a `.json` suffix?

Reply:
If we should only have a single staged plan, why don't we just use a fixed filename for it, like `loq-recovery-plans/staged.json`? Then we don't have to do all this filename checking and multi-plan detection and all of the other stuff, it becomes trivial to manage.

Reply:
The only benefit of having different file names is ability to get status without deserializing file or having a separate "manifest".
Since it is not a hot path of any sort and file shouldn't be large it is a reasonable trade off to have a single file and always load it to read the id. Makes code simpler indeed.